---
title: "Guinea worm mitochondrial DNA genomics - geospatial modeling "
output: html_notebook
---

## Load libraries, scripts, and data

```{r setup}
# load libraries
for(p in c('data.table', 'dplyr', 'tidyr', "ggplot2", "wesanderson")){
  if(!p %in% installed.packages()[,1]){
    install.packages(p)
    library(p, character.only = TRUE)
  }
  library(p, character.only = TRUE)
}

# set global options
options(datatable.fread.datatable=FALSE)
options(stringsAsFactors = FALSE)

# set directories
code_dir <- "/home/jribado/git/DDA-Genetics-GuineaWormTransmissionInChad/sanger_processing"
project_dir <- "/home/jribado/Dropbox (IDM)/Data, Dynamics, and Analytics Folder/Projects/Guinea Worm Genetics in Chad/gw_bwaAlign/2020_newSeqs"
# project_dir <- "/home/jribado/Dropbox (IDM)/PLoSNTDS_GuineaWormPhylodynamics/Figures"

# load functions
source(paste(code_dir, "plotting_themeJ.R", sep="/"))
source(paste(code_dir, "genSpatial_functions.R", sep="/"))
```

```{r load data}
barcode_vars <- read.delim(paste(project_dir, "202003_geneBarcodes.txt", sep="/"), sep="\t")
worm_meta <- read.delim(paste(project_dir, "202004_wormMetadata.txt", sep="/"), sep="\t")
merged_meta <- dplyr::left_join(barcode_vars, worm_meta) 
gw_comp <- data.table::fread(paste(project_dir, "202004_genePairwiseMetadata.txt", sep="/"), sep="\t") %>% unique()

gw_compFilt     <- read.delim(paste(project_dir, "20200515_bcPairwiseManualMeta.txt", sep="/"))
merged_metaFilt <- read.delim(paste(project_dir, "20200515_wormMeta.txt", sep="/"))

# remove outlier worms
# gps_out <- readRDS(paste(project_dir, "GPS_outlierHosts.Rds", sep="/")) 
# gw_compFilt <- dplyr::filter(gw_comp, !host_number.x %in% gps_out,  !host_number.y %in% gps_out)
# merged_metaFilt <- dplyr::filter(merged_meta, !host_number %in% gps_out)
```

```{r barcode_counts}
# identify common barcodes
gps_worms <-dplyr::filter(merged_metaFilt, !is.na(latitude) & !is.na(longitude))
  barcode_counts <- data.frame(table(gps_worms$full_barcode), stringsAsFactors = F) %>% 
  dplyr::arrange(desc(Freq)) %>%
  dplyr::rename("full_barcode" = Var1, count = Freq) %>%
  tibble::rowid_to_column("barcode_number")
# write.table(barcode_counts, paste(project_dir, "plots", "barcode_counts.txt", sep="/"), sep="\t", quote = F, row.names = F)
barcodes_keep <- dplyr::filter(barcode_counts, count >= 10) %>% .[["full_barcode"]]

barcode_tmp <- dplyr::inner_join(barcode_vars, barcode_counts)
# set colors for common barcodes
# bc_colors <- c(ggsci::pal_simpsons("springfield")(10), "grey50")
bc_colors <- c(ggsci::pal_futurama("planetexpress")(11)[-c(9, 11)], "grey50")
```


## Compare within host similarity

Some hosts had multiple worms. Is each worm within each host genetically identical?

```{r host_worms}
host_count <- dplyr::group_by(merged_metaFilt, year, host_number) %>%
  dplyr::summarise(host_count = n()) %>%
  dplyr::filter(host_count > 1) %>%
  dplyr::left_join(., merged_metaFilt) %>%
  dplyr::left_join(., barcode_tmp) %>%
  dplyr::mutate(barcode = ifelse(barcode_number > 9 | is.na(barcode_number), "Not common", barcode_number))
host_ord <- names(sort(table(host_count$host_number), decreasing = F))

host_bc_count <- host_count %>%
  dplyr::mutate(host_number = factor(host_number, levels = host_ord)) %>%
  ggplot(aes(x= host_number)) + 
  geom_bar(aes(fill=barcode)) +
  scale_fill_manual(values=c(bc_colors[-10], "golden rod"), name="Barcode identifier") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.border = element_rect(colour = "black", fill=NA, size=0.5),
        legend.position = c(0.28,0.95), 
        legend.box = "horizontal",
        panel.grid.major = element_blank()) +
  labs(x="Host", y="Worms per host") +
  guides(fill=guide_legend(ncol=5))
host_bc_count
ggsave("20200515_genSim_hostBarcode.png", plot = host_bc_count, path = paste(project_dir, "plots", sep="/"), width = 7.25, height = 3, units = c("in"), dpi = 300)
```

What is the genetic similarity of worms thats 


```{r host_sim_comp}
host_gen_sim <- dplyr::filter(gw_compFilt, dist_m == 0) %>%
  ggplot(aes(x=value, fill=host_match)) + 
  geom_histogram(position = 'identity', alpha = .75, bins=40) +
  scale_fill_manual(values=tf_colors, name="Same\nhost") +
  theme(panel.border = element_rect(colour = "black", fill=NA, size=0.5),
        legend.position = c(0.10,0.95), 
        legend.box = "horizontal",
        panel.grid.major = element_blank()) +
  labs(x="Genetic similarity", y="Pairwise comparisons")
host_gen_sim
ggsave("20200515_genSim_hostHistogram.png", plot = host_gen_sim, path = paste(project_dir, "plots", sep="/"), width = 7.25, height = 3, units = c("in"), dpi = 300)
```

```{r gen_sim_plot}
merge <-ggpubr::ggarrange(host_bc_count, host_gen_sim, ncol=1, labels=c("A", "B"))
ggsave("20200515_genSim_mergedHistogram.png", plot = merge, path = paste(project_dir, "plots", sep="/"), width = 7.25, height = 5, units = c("in"), dpi = 300)
```

### Emperical densities - Barcodes 

We can compare the densities for geneti similarity between matching barcodes and non-matching barcodes and distance. Since humans overlap with dogs in barcodes but are not 

```{r bc_match}
bcAll_df <- distDensity_plots(gw_compFilt, "20200515_bcAllFilt_density", project_dir)
table(dplyr::filter(gw_compFilt, !is.na(dist_m)) %>% .[["bc_match"]])
bcHum_df <- distDensity_plots( dplyr::filter(gw_compFilt, grepl("PD", worm1) & grepl("PD", worm2)), "20200515_bcHumFilt_density", project_dir)
bcDog_df <- distDensity_plots( dplyr::filter(gw_compFilt, !grepl("PD", worm1) & !grepl("PD", worm2)), "20200515_bcDogFilt_density", project_dir)
```

Do any barcodes seem to have a stronger effect than others on this curve? It doesn't make sense to include barcodes that are found in low frequency.


```{r ind_barcodes}
# subset "common" barcodes from the comparisons
comp_sub <- dplyr::filter(gw_compFilt, full_barcode.x %in% barcodes_keep, full_barcode.y %in% barcodes_keep) %>%
  dplyr::mutate(bc_group = ifelse(bc_match =="True", full_barcode.x, "No shared barcode in pair.")) %>%
  dplyr::left_join(., barcode_counts, by=c("bc_group" = "full_barcode")) %>%
  dplyr::mutate(Barcode = ifelse(!is.na(barcode_number), paste(barcode_number, "(N =", count, "worms)"), "No common barcode in pair."),
                Barcode = factor(Barcode, levels = gtools::mixedsort(unique(Barcode))))
table(dplyr::filter(comp_sub, !is.na(dist_m)) %>% .[["Barcode"]])

cdf <- ggplot(comp_sub, aes(x=dist_m/1000)) + 
  stat_ecdf(aes(colour = Barcode), size=1) +
  labs(x="Distance (km)",y="Cumulative density") +
  scale_color_manual(values= bc_colors, name="Barcode\nidentifier") 
pdf <- ggplot(comp_sub, aes(x=dist_m/1000)) + 
  stat_density(aes(fill = Barcode), position = "identity", alpha=0.75) +
  labs(x="Distance (km)", y="Probability density") +
  scale_fill_manual(values= bc_colors, name="Barcode\nidentifier") 
bc_df <- ggpubr::ggarrange(pdf, # + theme(axis.title=element_blank(), axis.text=element_blank(), plot.margin = unit(c(2,2,2,2), "lines")) , 
                           cdf, #+ theme(axis.title=element_blank(), axis.text=element_blank(), plot.margin = unit(c(2,2,2,2), "lines")), 
                           common.legend=TRUE, legend = "top")
bc_df
ggsave("20200515_bwa_bcIndFilt.png", plot = bc_df, path = paste(project_dir, "plots", sep="/"), width = 6, height = 3, units = c("in"), dpi = 300)
```

#### Sensitivity test - permuations 

```{r permuation}
dist_perm <- data.table(do.call(rbind, lapply(seq(1,100), function(i){
  tmp_corr <- dplyr::select(gw_compFilt, worm1, worm2, bc_match, dist_m)
  tmp_corr$dist_perm <- sample(tmp_corr$dist_m)
  tmp_corr$perm <- i
  return(tmp_corr)
})))

perm_plot <- ggpubr::ggarrange(
  ggplot(dist_perm, aes(x=dist_perm/1000, group=paste(bc_match, perm))) + 
    stat_ecdf(aes(colour = bc_match), size=1, alpha=0.25) +
    labs(x="Distance (km)", y="Cumulative density") +
    scale_color_manual(values=tf_colors, name="Same\nbarcode"),
  ggplot(dist_perm, aes(x=dist_perm/1000, group=paste(bc_match, perm))) + 
    stat_density(aes(fill = bc_match), position = "identity", alpha=0.25) +
    scale_fill_manual(values=tf_colors, name="Same\nbarcode") +
    labs(x="Distance (km)", y="Probability density"),
  common.legend=TRUE, legend = "top")
ggsave("20200515_sensitivity_densityDistPerm.png", plot = perm_plot, path = paste(project_dir, "plots", sep="/"), width = 6, height = 3, units = c("in"), dpi = 300)
```


#### Sensitivity test - single worm per host

We have a few worms that have multiple worms which can inflate the distance score, despite showing that hosts can carry multiple worms that are not genetically identical. How well do the distributions hold when hold when I randomly subsample one worm from each host?

```{r host_subsample}
host_perm <- data.table(do.call(rbind, lapply(seq(1,100), function(i){
  tmp <- dplyr::group_by(gps_worms, host_number) %>% dplyr::sample_n(1)
  tmp_corr <- dplyr::filter(gw_compFilt, worm1 %in% tmp$worm, worm2 %in% tmp$worm)
  tmp_corr$sample <- i
  return(tmp_corr)
})))

sub_samp <- ggpubr::ggarrange(
  ggplot(host_perm, aes(x=dist_m/1000, group=paste(bc_match, sample))) + 
    stat_ecdf(aes(colour = bc_match), size=1, alpha=0.25) +
    labs(x="Distance (km)", y="Cumulative density") +
    scale_color_manual(values=tf_colors, name="Same\nbarcode"),
  ggplot(host_perm, aes(x=dist_m/1000, group=paste(bc_match, sample))) + 
    stat_density(aes(fill = bc_match), position = "identity", alpha=0.25) +
    scale_fill_manual(values=tf_colors, name="Same\nbarcode") +
    labs(x="Distance (km)", y="Probability density"),
  common.legend=TRUE, legend = "top")
ggsave("20200515_sensitivity_densitySingHost.png", plot = sub_samp, path = paste(project_dir, "plots", sep="/"), width = 6, height = 3, units = c("in"), dpi = 300)
```

#### Sensitivity test - nucleotide diversity collapse

For a sensistivity test, what happens if we merge barcodes with a one nucleotide difference?

```{r barcode_sens}
sens_df <- dplyr::select(gw_compFilt, full_barcode.x, full_barcode.y) %>% unique() 
sens_diff <- do.call(rbind, lapply(1:nrow(sens_df), function(i){
  string_diff(sens_df$full_barcode.x[i], sens_df$full_barcode.y[i])})) %>% as.data.frame() %>%
  dplyr::left_join(., barcode_counts, by=c("full_barcode.x" = "full_barcode")) %>%
  dplyr::left_join(., barcode_counts, by=c("full_barcode.y" = "full_barcode"))
```

```{r sens_1off}
# identify barcode pairs that are one bp off
sens_1off <- dplyr::filter(sens_diff, !grepl("\\,|ignore", diff_pos)) %>% dplyr::select(full_barcode.x, full_barcode.y, diff_pos)
gw_compBC <- dplyr::left_join(gw_compFilt, sens_1off) %>%
  dplyr::mutate(bc_match = ifelse(!is.na(diff_pos), "True",  bc_match))
table(dplyr::filter(gw_compBC, !is.na(dist_m)) %>% .[["bc_match"]])
# create 
bc1off <- distDensity_plots(gw_compBC, "20200515_bcAllFilt_1off_density", project_dir)
```

```{r sens_2off}
# identify barcode pairs that are one bp off
sens_2off <- dplyr::filter(sens_diff, !grepl("ignore", diff_pos))  %>% dplyr::select(full_barcode.x, full_barcode.y, diff_pos)
gw_compBC <- dplyr::left_join(gw_compFilt, sens_2off) %>%
  dplyr::mutate(bc_match = ifelse(!is.na(diff_pos), "True",  bc_match))
table(gw_compBC$bc_match)
# create 
bc2off <- distDensity_plots(gw_compBC, "20200515_bcAllFilt_2off_density", project_dir)
```

The effect is dimished when we incorporate irgnoring two variants in a barcode, but not so much with one. Sensitivity analyses sugguest we can keep going.

For individual barcodes, it's not straight forward how to collapse them since certain barcodes (such as 6) share different one-off variant posiitons with other barcodes. It doesn't seem correct to collapse them knowing there may be up to 5 different variants in a group. Can revisit if reviewers suggest. 


### Rarefaction curves

For each worm, calculate the next closest worm and the barcode diversity.

```{r rarefaction_function}
gw_sub <- dplyr::select(gw_compFilt, worm1, full_barcode.x, worm2, full_barcode.y, dist_m) %>%
  dplyr::filter(!is.na(dist_m)) %>% unique()
dist_worms <- unique(c(gw_sub$worm1, gw_sub$worm2))

worm_rarefactionList <- lapply(setNames(dist_worms, dist_worms), function(worm){
  worm_sub <- dplyr::filter(gw_sub, worm1 == worm | worm2 == worm) %>% dplyr::arrange(dist_m)
  uniq_bc_counts <- bind_rows(lapply(1:nrow(worm_sub), function(i){
    cbind.data.frame(
      distance=worm_sub[i,]$dist_m, 
      step = i,
      unique_barcodes=length(unique(c(worm_sub[1:i,]$full_barcode.x, worm_sub[1:i,]$full_barcode.y)))-1)
  })) 
  uniq_bc_countsRm <- cbind.data.frame(worm = worm, 
                                       uniq_bc_counts,
                                       # dplyr::group_by(uniq_bc_counts, distance) %>% top_n(1, unique_barcodes), 
                                       stringsAsFactors=F)
})
worm_rarefaction <- unique(dplyr::bind_rows(worm_rarefactionList))
```

```{r rarefaction_plot}
barcode_number   <- dplyr::left_join(dplyr::select(barcode_vars, worm, full_barcode), barcode_counts)
worm_rarefaction <- dplyr::left_join(worm_rarefaction, barcode_number) %>% 
  dplyr::mutate(color = ifelse(full_barcode %in% barcodes_keep, paste(barcode_number, "\n(N =", count, "worms)"), "Not common"))
  
worm_rarefactionPlot <- worm_rarefaction %>% 
  dplyr::group_by(worm, distance, full_barcode, barcode_number) %>% 
  dplyr::mutate(color = ifelse(full_barcode %in% barcodes_keep, paste(barcode_number, "\n(N =", count, "worms)"), "Not common.\n(N = 54 worms)"),
                color = factor(color, levels = gtools::mixedsort(unique(color)))) %>%  
  ggplot(aes(x=distance/1000, y=unique_barcodes, group=worm)) +
  geom_line(aes(color=color), alpha=0.5, size=1) +
  labs(x="Distance (km)", y="Unique barcodes") +
  scale_color_manual(values=c(bc_colors[-10], "golden rod"), name="Barcode") 
# ggsave("bwa_wormRarefaction.png", plot = worm_rarefactionPlot, path = paste(project_dir, "plots", sep="/"), width = 6, height = 4, units = c("in"), dpi = 300)
ggsave("bwa_wormRarefactionInd.png", plot = worm_rarefactionPlot + facet_wrap(~color, ncol=4) + guides(color=F), 
        path = paste(project_dir, "plots", sep="/"), width = 6, height = 4, units = c("in"), dpi = 300)
rare_merged <-ggpubr::ggarrange(worm_rarefactionPlot + guides(color=F), 
                                worm_rarefactionPlot + facet_wrap(~color, ncol=4) + guides(color=F) +
                                  theme(axis.text   = element_text(size=rel(0.4), color="grey50"),
                                        strip.text  = element_text(face="bold", size=rel(.4)))
                                  )
#ggsave("bwa_wormRarefactionMerged.png", plot = rare_merged, path = paste(project_dir, "plots", sep="/"), width = 7.25, height = 4, units = c("in"), dpi = 300)
```

Follow-up, are these asymptotes driven by sampling scheme?

```{r}
sampling_dist <- worm_rarefaction %>% 
  dplyr::mutate(color = ifelse(full_barcode %in% barcodes_keep, paste(barcode_number, "\n(N =", count, "worms)"), "Not common.\n(N = 54 worms)"),
                color = factor(color, levels = gtools::mixedsort(unique(color)))) %>%   
  ggplot(aes(x=step, y=distance/1000, group = worm)) + 
  geom_line(aes(color=color), alpha=0.5, size=1) +
  scale_color_manual(values=c(bc_colors[-10], "golden rod"), name="Barcode", guide=F) +
  labs(x="Nearest worm step \n(max 370)", y="Distance (km)") +
  facet_wrap(~color, ncol=4)
ggsave("bwa_wormRarefactionDistances.png", plot = sampling_dist, path = paste(project_dir, "plots", sep="/"), width = 6, height = 4, units = c("in"), dpi = 300)
```

Yep, this suggests the asymptote for each barcode is a sampling scheme artifact. The consistent slope across the barcodes is still interesting since it does support that without genomics we would not know worms are linked since the cases are ontop of each other. 